lib/browser-specific.js

```diff
@@ -0,0 +1,247 @@
+'use strict';
+
+/**
+ * Implements functionality to report on how many WPT tests fail only on one
+ * browser (aka 'browser-specific failures).
+ */
+
+// A helper class providing an iterator-like interface to either an Object with
+// enumerable properties, or an Array. Iterates in a sorted order; alphabetical
+// in the case of an Object, or in the order defined by |comparatorFunc| in the
+// case of an Array.
+class IteratorHelper {
+  constructor(arrOrObject, comparatorFunc = null) {
+    this.currentIndex = 0;
+    this.values = arrOrObject;
+
+    if (this.values instanceof Array) {
+      this.keys = null;
+      this.maxIndex = this.values.length;
+      this.values.sort(comparatorFunc);
+    } else {
+      this.keys = Object.keys(this.values);
+      this.maxIndex = this.keys.length;
+      this.keys.sort();
+    }
+  }
+
+  hasNext() {
+    return this.currentIndex < this.maxIndex;
+  }
+
+  next() {
+    if (!this.hasNext()) {
+      throw new Error('Cannot move next; out of values');
+    }
+    this.currentIndex++;
+  }
+
+  currentKey() {
+    if (this.keys === null) {
+      throw new Error('Cannot get key of an Array iteration');
+    }
+    return this.keys[this.currentIndex];
+  }
+
+  currentValue() {
+    if (this.keys === null) {
+      return this.values[this.currentIndex];
+    }
+    return this.values[this.currentKey()];
+  }
+}
+
+// Scores a particular WPT test for a set of browsers, updating |scores|. It is
+// assumed that the order of |browserTests| and |scores| is the same.
+function scoreTest(browserTests, scores) {
+  // Some WPT tests contain multiple 'subtests' (e.g. most jsharness tests),
+  // whilst others are just a single conceptual test (e.g. reftests).
+  //
+  // Tests without subtests are scored as a simple 0-or-1 for each failing
+  // browser (0 if any other browser also fails, 1 if no other browser fails).
+  // When there are subtests, we do a similar calculation per-subtest, but
+  // normalize the results by the number of subtests in the test. This stops
+  // tests with thousands of subtests from dominating the results.
+  if (browserTests.every(tests => !tests.subtests || tests.subtests.length == 0)) {
+    let failed = [];
+    for (let i = 0; i < browserTests.length; i++) {
+      // TODO: Determine exact criteria for a BSF at a test level.
+      if (browserTests[i].status !== 'PASS') {
+        failed.push(i);
+      }
+    }
+
+    if (failed.length == 1) {
+      scores[failed[0]] += 1;
+    }
+  } else if (browserTests.every(t => t.subtests && t.subtests.length > 0)) {
+    const comparator = (s1, s2) => (s1.name > s2.name) ? 1 : -1;
```

Comment:
Should the comparator return 0 if `s1.name === s2.name`? Consider using `s1.name.localeCompare(s2.name)`, it actually won't matter if the local is different since this only point is to line up the names that are equal.